zkevm-circuits/src/circuit_tools/memory.rs

```diff
@@ -1,47 +1,64 @@
 //! Memory
-use crate::util::{query_expression, Expr};
+use crate::{
+    evm_circuit::util::rlc,
+    util::{query_expression, Expr},
+};
 use eth_types::Field;
 use halo2_proofs::{
     circuit::Value,
-    plonk::{Advice, Column, ConstraintSystem, Error, Expression},
+    plonk::{
+        Advice, Column, ConstraintSystem, Error, Expression, FirstPhase, SecondPhase, ThirdPhase,
+    },
     poly::Rotation,
 };
-use std::ops::{Index, IndexMut};
+use itertools::Itertools;
+use std::{
+    collections::HashMap,
+    ops::{Index, IndexMut},
+};
 
 use super::{
-    cached_region::CachedRegion, cell_manager::CellType, constraint_builder::ConstraintBuilder,
+    cached_region::CachedRegion,
+    cell_manager::{Cell, CellType},
+    constraint_builder::ConstraintBuilder,
 };
 
 #[derive(Clone, Debug, Default)]
 pub(crate) struct Memory<F, C> {
-    columns: Vec<Column<Advice>>,
+    height: usize,
     banks: Vec<MemoryBank<F, C>>,
+    rw_records: HashMap<C, (Column<Advice>, Column<Advice>)>,
 }
 
 impl<F: Field, C: CellType> Memory<F, C> {
-    pub(crate) fn new(columns: Vec<Column<Advice>>) -> Self {
+    pub(crate) fn new(
+        meta: &mut ConstraintSystem<F>,
+        tags: Vec<(C, usize)>,
+        offset: usize,
+        height: usize,
+    ) -> Self {
+        let mut rw_records = HashMap::new();
+        let banks = tags
+            .iter()
+            .map(|(tag, phase)| {
+                let [key, reads, writes] = match phase {
+                    1 => [(); 3].map(|_| meta.advice_column_in(FirstPhase)),
+                    2 => [(); 3].map(|_| meta.advice_column_in(SecondPhase)),
+                    3 => [(); 3].map(|_| meta.advice_column_in(ThirdPhase)),
+                    _ => unreachable!(),
+                };
```

Comment:
Same phase logic already happens in `CellConfig` which should get reused somehow.

Reply:
Still duplicated here.

Reply:
Yea, i was gonna move it back last time but i went ahead to refactor the lookup system itself, and I couldn't use the API before finishing that (i.e. `add_lookup()` now instead of 'add_dynamic_lookup()' in the last version). So that's done now, refactoring this would be easy :)